### PR TITLE
query API knows when relationship is encrypted

### LINKED ISSUE #43: How can we tell the difference between public and private follows/blocks?

### PR DESCRIPTION
**Pull request by staltz**
This fixes #43 

- [x] `db2Contacts.js2` updated
- [x] `contacts.js` updated
- [x] `isFollowing` updated to add `opts.details`
- [x] `isBlocking` updated to add `opts.details`
- [x] tests
- [x] https://github.com/ssbc/ssb-friends/pull/71 merged first
- [x] Add test that restarting sbot still works
- [x] Fix `contact.js` loading part
- [x] Check the storage cost of the additional `privately` boolean

### CODE REVIEW

**Comment by staltz** [comment]
>  Check the storage cost of the additional privately boolean

I wanted to check what is the overhead of changing the index value from `edgeValue` to `[edgeValue, privately]` so I tried some alternatives. My `@QlCTpv` `~/.ssb/db2/indexes/contacts` folder had

| | Size in bytes |
|--|----|
| `edgeValue` (master branch) | 6 347 000 |
| `[edgeValue, privately]` | 9 155 062 |
| `[edgeValue, true]` or `[edgeValue]` | 8 048 349 |
| `[edgeValue, 1]` or `[edgeValue]` | 8 048 275 |
| `"a" + edgeValue` or `"b" + edgeValue` | 8 223 268 |
| `"p" + edgeValue` or `edgeValue` | 6 896 162 |
| `10 + edgeValue` or `edgeValue` | 6 896 153 |

The +10 number hack is the smallest but has the obvious downside of limiting the largest edge distance to 9.

My favorite compromise between small size and low-WTF-effect is `"p" + edgeValue` & `edgeValue`. Thoughts, @arj03?

**Comment by arj03** [comment]
Agree, `p` seems like the best choice. It seems this (private/public) is actually a good case for keeping layers around. I was keen on removing it before this PR :-)

**Comment by staltz** [comment]
> Agree, `p` seems like the best choice. It seems this (private/public) is actually a good case for keeping layers around. I was keen on removing it before this PR :-)

Yeah! I was pleasantly surprised that it was a neat solution for this use case. :)

**Comment by staltz** [comment]
Hey @Powersource, feedback I'm looking from you is whether the README (the shape of the API) and tests look good.

@mixmix overall this whole PR, but I suspect you want to take a close look at `contacts.js`, the index for flume. While @arj03 could focus on the `db2Contacts.js` changes.

File: README.md [hunk 1]
```diff
@@ -80,16 +80,24 @@ publish this contact message privately to some feeds / groups (see e.g.
...
 `opts.source` and `opts.dest` are strings of SSB Feed IDs.
 
-### `ssb.friends.isBlocking({source, dest}, cb)` ("async" muxrpc API)
+If you pass `opts.details = true`, then the callback will respond with the
+object `{ response, private }`, where `response` is the boolean indicating
+the follow relationship, and `private` is a boolean indicating that the
+relationship was originally encoded in a private (encrypted) message.
 
-Calls back `true` if `opts.source` blocks `opts.dest`, `false` otherwise.
+### `ssb.friends.isBlocking({source, dest}, cb)` ("async" muxrpc API)
```
Comment (Powersource) [inline README.md:91]: Shouldn't this say `{source, dest, details}` ? Or just `opts` like we seem to do for all the other function docs here.

Comment (staltz) [reply README.md:91]: Will fix!

File: test/privately-db2.js [hunk 2]
```diff
@@ -10,50 +10,93 @@ const mkdirp = require('mkdirp')
...
 
+  const [err4, details4] = await run(sbot.friends.isFollowing)({
+    source,
+    dest,
+    details: true
+  })
+  t.error(err4, 'no error')
+  t.deepEqual(
+    details4,
+    { response: true, private: true },
```
Comment (Powersource) [inline test/privately-db2.js:67]: Maybe also add assertions before the follows and blocks, where `details: true` but `response: false`

**Comment by Powersource** [comment]
other than that lgtm

**Review by arj03 — APPROVED** [review]
Looks good to me

**Pull request merged**